File: sdks/c++/basys.sdk.cc/tests/support/vab/stub/elements/SimpleVABElement.cpp
#include "SimpleVABElement.h"

#include <basyx/object.h>
#include <basyx/types.h>

#include <set>

namespace basyx {
namespace tests {
namespace support {

	class testFuncs {
	public:
		static int add(int a, int b) {
			return a + b;
		}
	};


	basyx::object make_simple_vab_element()
	{
		basyx::object vabElement = basyx::object::make_map();

		// Add primivites
		vabElement.insertKey("primitives", basyx::object::make_map());
		vabElement.getProperty("primitives").insertKey("integer", 123);
		vabElement.getProperty("primitives").insertKey("double", 3.14);
		vabElement.getProperty("primitives").insertKey("string", std::string("TestValue"));

		// Add function types
		vabElement.insertKey("operations", basyx::object::make_map());
		vabElement.getProperty("operations").insertKey("supplier", basyx::object::make_function([]() {return true; }));
		vabElement.getProperty("operations").insertKey("complex", basyx::object::make_function(testFuncs::add));
		vabElement.getProperty("operations").insertKey("serializable", basyx::object::make_function(testFuncs::add));

		vabElement.getProperty("operations").insertKey("invalid", true);
		vabElement.getProperty("operations").insertKey("invokable", basyx::object::make_function([]() {return true; }));

		// Add structure types
		vabElement.insertKey("structure", basyx::object::make_map());
		vabElement.getProperty("structure").insertKey("map", basyx::object::make_map());
		vabElement.getProperty("structure").insertKey("set", basyx::object::make_set<bool>());
		vabElement.getProperty("structure").insertKey("list", basyx::object::make_list<int>());

		// Add corner cases
		vabElement.insertKey("special", basyx::object::make_map());
		vabElement.getProperty("special").insertKey("casesensitivity", true);
		vabElement.getProperty("special").insertKey("caseSensitivity", false);
		vabElement.getProperty("special").insertKey("nested", basyx::object::make_map());
		vabElement.getProperty("special").getProperty("nested").insertKey("nested", basyx::object::make_map());
		vabElement.getProperty("special").getProperty("nested").getProperty("nested").insertKey("value", 100);
		vabElement.getProperty("special").insertKey("null", basyx::object::make_null());
		
		return vabElement;
	}

}
}
}

File: sdks/java/basys.sdk/src/test/java/org/eclipse/basyx/testsuite/regression/vab/modelprovider/SimpleVABElement.java
package org.eclipse.basyx.testsuite.regression.vab.modelprovider;

import java.io.Serializable;
import java.util.ArrayList;
import java.util.HashMap;
import java.util.HashSet;
import java.util.function.Function;
import java.util.function.Supplier;

import org.eclipse.basyx.vab.exception.ServerException;

/**
 * A simple VAB model that explains the use of the VAB primitives
 * 
 * @author kuhn, espen
 *
 */
public class SimpleVABElement extends HashMap<String, Object> {
	private static final long serialVersionUID = 3942399852711325850L;

	/**
	 * Constructor for a simple VAB element that contains all data types
	 */
	public SimpleVABElement() {
		// Add primitive types
		HashMap<String, Object> primitives = new HashMap<>();
		primitives.put("integer", 123);
		primitives.put("double", 3.14d);
		primitives.put("string", "TestValue");
		put("primitives", primitives);

		// Add function types
		HashMap<String, Object> functions = new HashMap<>();
		functions.put("supplier", (Supplier<Object>) () -> {
			return true;
		});
		functions.put("serverException", (Function<Object[], Object>) (param) -> {
			throw new ServerException("testExceptionType", "Exception description");
		});
		functions.put("nullException", (Function<Object[], Object>) (param) -> {
			throw new NullPointerException();
		});
		functions.put("complex", (Function<Object[], Object>) (param) -> {
			return (int) param[0] + (int) param[1];
		});
		functions.put("serializable", (Function<Object[], Object> & Serializable) (param) -> {
			return (int) param[0] + (int) param[1];
		});
		functions.put("invalid", true);
		functions.put("invokable", (Function<Object[], Object>) (param) -> {
			return true;
		});
		put("operations", functions);

		// Add structure types
		HashMap<String, Object> structure = new HashMap<>();
		structure.put("map", new HashMap<String, Object>());
		structure.put("set", new HashSet<Object>());
		structure.put("list", new ArrayList<Object>());
		put("structure", structure);

		// Add corner cases
		HashMap<String, Object> special = new HashMap<>();
		special.put("casesensitivity", true);
		special.put("caseSensitivity", false);
		HashMap<String, Object> nestedA = new HashMap<>();
		HashMap<String, Object> nestedB = new HashMap<>();
		nestedA.put("nested", nestedB);
		nestedB.put("value", 100);
		special.put("nested", nestedA);
		special.put("null", null);
		put("special", special);
	}
}
